### PR TITLE
Stackdriver template query editor

### LINKED ISSUE #13561: Stackdriver: Template query editor
The new Stackdriver datasource does not yet have support for template queries. The current approach for datasources without a way to create lists of values is to invent a function (like the [label_values](http://docs.grafana.org/features/datasources/prometheus/#query-variable) function for Prometheus).

For Stackdriver, we would like to improve Grafana and add support for a template query editor plugin so that each datasource can have its own custom template query editor. Stackdriver would be the first implementation.

The query editor should be able to create the following queries:

- ~~list of services~~
- list of metric types
- list of metric labels for a chosen metric type
- list of resource labels for a chosen metric type
- list of aggregations for a chosen metric type
- list of aligners for a chosen metric type
- list of alignment periods
- list of resource types

ref #13289

### PR DESCRIPTION
**Pull request by sunker**
Makes it possible for plugins to implement its own template query editor. If the module.js|ts file in the plugin folder does not export a `TemplateQueryComponent`, a default template query editor will be used. This default editor only contains a text field for the query - like the existing query editor. However if the plugin does have an export named `TemplateQueryComponent`, that will be used instead. 

If the plugin exports a TemplateQueryComponent, it needs to be a react component. The component will receive the template variable query data (if the variable was saved before) from the `query` property in the props object. The way the query is propagated back to the grafana framework is by using the `onChange` function that is defined in the component props. 

This PR not only provides functionality for loading the template query editor for a datasource, but also the implementation of a template query editor for the Stackdriver datasource.  More on that [here](https://github.com/grafana/grafana/issues/13561)

closes #13561

### CODE REVIEW

File: public/app/plugins/datasource/stackdriver/components/SimpleSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import React, { SFC } from 'react';
+
+interface Props {
+  onValueChange: (e) => void;
+  options: any[];
+  value: string;
+  label: string;
+}
+
+const SimpleSelect: SFC<Props> = props => {
```
Comment (torkelo) [inline public/app/plugins/datasource/stackdriver/components/SimpleSelect.tsx]: Think we already have a similar component, https://github.com/grafana/grafana/blob/master/public/app/core/components/Picker/SimplePicker.tsx

File: public/app/types/plugins.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ export interface PluginExports {
   QueryCtrl?: any;
   ConfigCtrl?: any;
   AnnotationsQueryCtrl?: any;
+  TemplateQueryComponent?: any;
```
Comment (torkelo) [inline public/app/types/plugins.ts]: think VariableQueryEditor is a better name.

File: public/app/features/templating/DefaultTemplateQueryComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import React, { PureComponent } from 'react';
+import { TemplateQueryProps } from 'app/types/plugins';
+
+export default class DefaultTemplateQueryComponent extends PureComponent<TemplateQueryProps, any> {
+  constructor(props) {
+    super(props);
+    this.state = { value: props.query };
+    this.handleChange = this.handleChange.bind(this);
+    this.handleBlur = this.handleBlur.bind(this);
```
Comment (peterholmberg) [inline public/app/features/templating/DefaultTemplateQueryComponent.tsx]: We've used the "on" prefix and written these as arrow functions. That way we don't have to bind them in the constructor.

**Review by peterholmberg — CHANGES_REQUESTED** [review]

File: public/app/plugins/datasource/stackdriver/components/SimpleSelect.tsx (hunk 1 above)
Comment (sunker) [reply public/app/plugins/datasource/stackdriver/components/SimpleSelect.tsx]: The only problem that I can see with using this component is that only a part of the whole page, in this case the stackdriver query editor, would get the look and feel of the new select option, whereas the rest of the page has the look and feel of the old angularjs select option. Wouldn't it be better to use the old style of the select until the whole template page is migrated to react?

File: public/app/features/templating/DefaultTemplateQueryComponent.tsx (hunk 3 above)
Comment (sunker) [reply public/app/features/templating/DefaultTemplateQueryComponent.tsx]: Fixed

File: public/app/types/plugins.ts (hunk 2 above)
Comment (sunker) [reply public/app/types/plugins.ts]: Fixed

**Review by daniellee — APPROVED** [review]
Couple of improvements to make like supporting multi value template variables for filtering and group by.

**Pull request merged**